source/mesh_refinement/isosurfaces.cc

```diff
@@ -0,0 +1,391 @@
+                AssertThrow(tmpNumber.size() == 2,
+                            ExcMessage("Could not convert value '" + string + "' to an int because it contains more than one '-' sign."));
+                return maximum_refinement_level-Utilities::string_to_int(tmpNumber[1]);
+              }
+            else
+              {
+                AssertThrow(string.compare(0,4,"maxi") != 0,
+                            ExcMessage("A value of " + string_value + " was provided, but you can't provide a larger value than the maximum."));
+                return maximum_refinement_level;
+              }
+          }
+        else
+          {
+            return Utilities::string_to_int(string);
+          }
+
+      }
+    }
+
+    template <int dim>
+    void
+    Isosurfaces<dim>::update ()
+    { }
+
+    template <int dim>
+    void
+    Isosurfaces<dim>::tag_additional_cells () const
+    {
+      if (this->get_dof_handler().n_locally_owned_dofs() == 0)
+        return;
+
+      const Quadrature<dim> quadrature(this->get_fe().base_element(this->introspection().base_elements.temperature).get_unit_support_points());
+      FEValues<dim> fe_values (this->get_mapping(),
+                               this->get_fe(),
+                               quadrature,
+                               update_quadrature_points | update_values | update_gradients);
+
+
+      MaterialModel::MaterialModelInputs<dim> in(quadrature.size(),
+                                                 this->n_compositional_fields());
+
+      unsigned i = 0;
+      for (const auto &cell : this->get_dof_handler().active_cell_iterators())
+        {
+          if (cell->is_locally_owned())
+            {
+              bool coarsen = false;
+              bool clear_refine = false;
+              bool refine = false;
+              bool clear_coarsen = false;
+
+              fe_values.reinit(cell);
+              in.reinit(fe_values, cell, this->introspection(), this->get_solution(), true);
+
+              for (unsigned int i_quad=0; i_quad<quadrature.size(); ++i_quad)
+                {
+                  for (auto &isosurface : isosurfaces)
+                    {
+                      // setup the vector to check
+                      std::vector<double> values(isosurface.min_values.size());
+                      for (unsigned int index = 0; index < isosurface.properties.size(); ++index)
+                        {
+                          if (isosurface.properties[index].name == Internal::PropertyName::Temperature)
+                            {
+                              values[index] = in.temperature[i_quad];
+                            }
+                          else if (isosurface.properties[index].name == Internal::PropertyName::Composition)
+                            {
+                              values[index] = std::min(std::max(in.composition[i_quad][isosurface.properties[index].index], 0.0), 1.0);
```

Comment:
Why are compositions assumed to be between 0. and 1.?

Reply:
I see no reason to restrict that here, but I assume this is based on what is done in `parse_map_to_double_array`. Is that correct @lhy11009?

Reply:
I think @tjhei 's comments make sense. But I just copy from the 'volume_fraction' function from the MaterialUtility class. Right now I am not sure what to do. Let discuss it.

Reply:
This is still an open problem.

Reply:
ah, sorry, I missed it somehow. I have just removed the bounds checks.